Builder/AbstractFormContractor.php

```diff
@@ -53,97 +56,127 @@ public function __construct(FormFactoryInterface $formFactory)
-            'sonata_type_model_autocomplete',
-            'sonata_type_collection',
-        ), true)) {
-            if (!$fieldDescription->getAssociationAdmin()) {
-                throw $this->createMissingAssociationAdminException($fieldDescription);
-            }
-        }
+        $this->ensureFormTypeCanHandleFieldDescription($fqcn, $fieldDescription);
 
-        if (in_array($type, array(
-            'sonata_type_model', 'sonata_type_model_list',
-            'sonata_type_model_hidden', 'sonata_type_model_autocomplete',
-        ), true)) {
-            $options['class'] = $fieldDescription->getTargetEntity();
-            $options['model_manager'] = $fieldDescription->getAdmin()->getModelManager();
-        } elseif ($type === 'sonata_type_admin') {
-            if (!$fieldDescription->describesSingleValuedAssociation()) {
-                throw new \LogicException(sprintf(
-                    'The `sonata_type_admin` type only handles single-valued associations.'
-                    .' Try using `sonata_type_collection` or one of the `sonata_type_model` variants'
-                    .' for field `%s`.',
-                    $fieldDescription->getName()
-                ));
-            }
+        $options = array(
+            'sonata_field_description' => $fieldDescription,
+        );
 
-            // set sensitive default value to have a component working fine out of the box
-            $options['delete'] = false;
-
-            $options['data_class'] = $fieldDescription->getAssociationAdmin()->getClass();
-            $fieldDescription->setOption('edit', $fieldDescription->getOption('edit', 'admin'));
-        } elseif ($type === 'sonata_type_collection') {
-            if (!$fieldDescription->describesCollectionValuedAssociation()) {
-                throw new \LogicException(sprintf(
-                    'The `sonata_type_collection` type only handles collection-valued associations.'
-                    .' Try using `sonata_type_admin` or one of the `sonata_type_model` variants'
-                    .' for field `%s`',
-                    $fieldDescription->getName()
-                ));
-            }
-            $options['type'] = 'sonata_type_admin';
-            $options['modifiable'] = true;
-            $options['type_options'] = array(
-                'sonata_field_description' => $fieldDescription,
-                'data_class' => $fieldDescription->getAssociationAdmin()->getClass(),
-            );
+        switch ($fqcn) {
+            case 'Sonata\AdminBundle\Form\Type\ModelType':
+            case 'Sonata\AdminBundle\Form\Type\ModelAutocompleteType':
+                if ($fieldDescription->describesCollectionValuedAssociation()) {
+                    $options['multiple'] = true;
+                }
+                $options['class'] = $fieldDescription->getTargetEntity();
+                $options['model_manager'] = $fieldDescription->getAdmin()->getModelManager();
+                break;
```

Comment:
I think lines 83 through 85 could be removed (with maybe a comment like `// no break here, this is intentional`)